### PR TITLE
Auth token support + kind cluster helper

### PR DESCRIPTION
**Pull request by mjudeikis**
In this PR:
1. Rename `readme.md` to `README.md` (sorry, nitpicking)
2. Add the ability to deploy with a token file
3. Remove lets encrypt production issuer as default (does not work well with kind)
4. Add helper script to deploy end-to-end kind cluster with self-signed certificates and localhost ingress (nginx)

### CODE REVIEW

File: charts/kcp/values.yaml [hunk 1]
```diff
@@ -43,15 +49,7 @@ kcpFrontProxy:
...
-    issuerSpec:
-      acme:
-        server: https://acme-v02.api.letsencrypt.org/directory
-        privateKeySecretRef:
-          name: kcp-front-proxy-issuer-account-key
-        solvers:
-        - http01:
-            ingress:
-              serviceType: ClusterIP
+    issuerSpec: {selfSigned: {}}
```
Comment (mjudeikis) [inline charts/kcp/values.yaml:52]: This part is complicated. As this is treated like a map, helm merges values here. 
I think the solution would be to keep this empty and ask users explicitly provide value via `values.yaml`
Having default in `values.yaml` as its currently is makes it impossible to override. 

It might be possible to workaround this somehow, but I didn't find a quick way. Need more experimentation .

**Comment by mjudeikis** [comment]
happy to "downsize" this PR if people think this is too much :)

File: README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
+# KCP Helm Charts
```
Comment (hardys) [inline README.md:1]: nit, if you rename with `git mv` I think the diff will only show the rename+changes vs a whole new file?

File: charts/kcp/templates/kcp.yaml [hunk 3]
```diff
@@ -323,6 +326,10 @@ spec:
           mountPath: /etc/kcp/tls/requestheader-client
         - name: kubeconfig
           mountPath: /etc/kcp/config
+        {{- if .Values.kcp.tokenAuth.enabled }}
+        - name: kcp-token-auth-file
+          mountPath: /etc/kcp/token-auth
+        {{- end}}
```
Comment (hardys) [inline charts/kcp/templates/kcp.yaml:332]: nit - would be nice to break the change into logically independent commits, e.g enabling token auth isn't strictly related to the KIND helper scripts

File: hack/kind-setup.sh [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+if [ ! -f "/usr/local/bin/kind" ]; then
+ echo "Installing KIND"
+ curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.14.0/kind-linux-amd64
+ chmod +x ./kind
+ sudo mv ./kind /usr/local/bin/kind
+else
+    echo "KIND already installed"
+fi
+
+CLUSTER_NAME=kcp
```
Comment (hardys) [inline hack/kind-setup.sh:12]: Perhaps make this user-modifiable vs hard-coded?, e.g 

`CLUSTER_NAME=${CLUSTER_NAME:-kcp}`

File: README.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+          enabled: true
+          annotations:
+            kubernetes.io/ingress.class: "nginx"
+            nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
+      ```
+
+# Known issues
+
+* https://github.com/kcp-dev/kcp/issues/2295 - Deployments fail to start.
+Workaround: Delete the empty token file in `kcp` volume and restart kcp pod.
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply README.md:100]: This happens almost always in my kind cluster on new deployments :/ 
@s-urbaniak did you manage to get to the bottom of this? If not we can look into my setup as it almost always replicatable

[the remark this answers is not in the document]
Comment (ncdc) [reply README.md:100]: I got to the bottom of it. You can't delete .kcp/admin.kubeconfig but preserve .kcp/.admin-token-store. We used to do that in kcp CI make targets but https://github.com/kcp-dev/kcp/pull/2707/files fixed it.

File: hack/generate-admin-kubeconfig.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+hostname=$(cat ./hack/kcp/values.yaml | grep externalHostname | cut -d" " -f2- | tr -d '"')
```
Comment (hardys) [inline hack/generate-admin-kubeconfig.sh:3]: perhaps it might be cleaner to do `--set externalHostname=$EXTERNAL_HOSTNAME` in kind-setup.sh, then export that variable for use here?

Or else this could be inlined in the script, if we don't expect it to be called outside the script?

File: hack/kcp/values.yaml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+externalHostname: "kcp.dev.local"
+kcp:
+  volumeClassName: "standard"
+  tokenAuth:
+    enabled: true
```
Comment (hardys) [inline hack/kcp/values.yaml:5]: perhaps this should be `hack/kind-values.yaml`?

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (hardys) [reply README.md:100]: > I got to the bottom of it. You can't delete .kcp/admin.kubeconfig but preserve .kcp/.admin-token-store. We used to do that in kcp CI make targets but https://github.com/kcp-dev/kcp/pull/2707/files fixed it.

@ncdc I'm still hitting this issue (I had to do the workaround today using the latest image from https://github.com/kcp-dev/kcp/pkgs/container/kcp) but https://github.com/kcp-dev/kcp/pull/2707 looks like it merged 2 weeks ago?

[the remark this answers is not in the document]
Comment (hardys) [reply README.md:100]: I think it may be around [here](https://github.com/kcp-dev/kcp/blob/main/pkg/server/options/authentication.go#L99) - if the token hash file exists but the admin kubeconfig doesn't, we'll end up with an empty shardAdminToken?

**Comment by hardys** [comment]
> happy to "downsize" this PR if people think this is too much :)

Personally I think it's OK, but it would be nice to break it into a few commits if possible since it's not all strictly related to the kind helper script.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply README.md:100]: Yes, that's what I meant by you can't delete the admin.kubeconfig file by itself.

File: README.md (hunk 2 above)
Comment (mjudeikis) [reply README.md:1]: `git mv readme.md README.md` :/ didn't worked

Comment (hardys) [reply README.md:1]: In what way didn't work?  I did it [here] (using git 2.39.1 on Fedora) and worked fine :thinking: 

Given the small amount of history on the file it's not a big deal, but would be interesting to understand what the problem is :)

Comment (mjudeikis) [reply README.md:1]: same git, same OS... Honestly, it never worked on my git config :D I might need to look into which githook breaks this behaviour on my ned :D

**Comment by mjudeikis** [comment]
@hardys  is there anything else you want me to change on this?

File: README.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                  serviceType: ClusterIP
+    ```
+### Install to KIND cluster (for development)
+
+There is a helper script to install KCP to a KIND cluster. It will install cert-manager, nginx-ingress and KCP.
+Kind cluster binds to host ports 80 and 443 for ingress. Ingress is emulated using host entries in `/etc/hosts`.
+This particular configuration is useful for development and testing, but will not work with LetsEncrypt.
+
+    ```bash
+    ./hack/kind-setup.sh
```
Comment (mamachanko) [inline README.md:78]: This fails for me like so:

```
❯ ./hack/kind-setup.sh
KIND already installed
No kind clusters found.
Creating cluster "kcp" ...
 ✓ Ensuring node image (kindest/node:v1.24.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kcp"
You can now use your cluster with:

kubectl cluster-info --context kind-kcp --kubeconfig ./kcp.kubeconfig

Thanks for using kind! 😊
Installing ingress
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
Error from server (NotFound): nodes "faros-control-plane" not found
```

Comment (mamachanko) [reply README.md:78]: With this fix:
```patch
diff --git a/hack/kind-setup.sh b/hack/kind-setup.sh
index 93fd8f1..ac179ad 100755
--- a/hack/kind-setup.sh
+++ b/hack/kind-setup.sh
@@ -25,7 +25,7 @@ export KUBECONFIG=./${CLUSTER_NAME}.kubeconfig
 echo "Installing ingress"
 
 kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/kind/deploy.yaml
-kubectl label nodes faros-control-plane node-role.kubernetes.io/control-plane-
+kubectl label nodes kcp-control-plane node-role.kubernetes.io/control-plane-
 
 echo "Waiting for the ingress controller to become ready..."
 kubectl --context "${KUBECTL_CONTEXT}" -n ingress-nginx wait --for=condition=Ready pod -l app.kubernetes.io/component=controller --timeout=5m
```
I get a little further, but it stills fails eventually:
```
❯ ./hack/kind-setup.sh
KIND already installed
No kind clusters found.
Creating cluster "kcp" ...
 ✓ Ensuring node image (kindest/node:v1.24.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kcp"
You can now use your cluster with:

kubectl cluster-info --context kind-kcp --kubeconfig ./kcp.kubeconfig

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/
Installing ingress
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
node/kcp-control-plane unlabeled
Waiting for the ingress controller to become ready...
error: no matching resources found

```

Comment (mamachanko) [reply README.md:78]: Still:
```
❯ git rev-parse @
58d8622466f304bc02ee7fdcaef39b45f8c7df09

kcp-helm-charts  local.dev [$]  ⊙ 4

❯ ./hack/kind-setup.sh
KIND already installed
No kind clusters found.
Creating cluster "kcp" ...
 ✓ Ensuring node image (kindest/node:v1.24.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kcp"
You can now use your cluster with:

kubectl cluster-info --context kind-kcp --kubeconfig ./kcp.kubeconfig

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
Installing ingress
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
node/kcp-control-plane unlabeled
Waiting for the ingress controller to become ready...
error: no matching resources found
```

Comment (mjudeikis) [reply README.md:78]: This implies ingress is not deploying for you. Can you check the cluster to see what is happening there?

Comment (mamachanko) [reply README.md:78]: `ingress-nginx` is deploying fine, but there's a race condition.

The installation of `ingress-nginx` and waiting for its readiness happens immediately after the kind cluster is created. At that time the control plane-related pods within the cluster haven't even started yet. As a result, at that moment it seems as if `ingress-nginx` is not installing.

If you bump the sleep to `30` it works, because everything had a chance to wake up. At least on my machine. `sleep 10` was too soon as well.

I don't think that sleeping here is the right thing. Maybe decouple cluster creation and initialization into different scripts. Or, better, find a more resilient and idiomatic approach to waiting.

Comment (mjudeikis) [reply README.md:78]: I pushed some different changes. 
Moving to a separate script IMHO will not solve anything as we still have the same issue, just in the other script.

Comment (mamachanko) [reply README.md:78]: I'm afk for some days. I've seen the changes you and assume it's working. Thanks for heeding my feedback!

**Comment by mjudeikis** [comment]
@hardys whats missing to merge this?

**Comment by xrstf** [comment]
Closing to re-open again, just to get Prow working. Please do not be alarmed.

File: README.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# KCP Helm Charts
+ KCP Helm Charts
```
Comment (xrstf) [inline README.md:1]: Why no heading anymore?

File: hack/generate-admin-kubeconfig.sh [hunk 10]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+set -e
+
+hostname=$(cat ./hack/kind-values.yaml | grep externalHostname | cut -d" " -f2- | tr -d '"')
```
Comment (xrstf) [inline hack/generate-admin-kubeconfig.sh:4]: Couldn't we use `yq` (https://github.com/mikefarah/yq) for this, rather than grepping through YAML?

Comment (mjudeikis) [reply hack/generate-admin-kubeconfig.sh:4]: I don't mind changing it. My general "rule of thumb" is if tools needs to be installed and is not part of most linux OS - its should not be in the readme, if explicitly not specified that its pre-requisite. Makes it cleaner
